addons/crm/models/crm_lead.py

```diff
@@ -930,15 +931,91 @@ def action_schedule_meeting(self):
+            Allday events datetimes are used without taking tz into account.
+            -If there is no event, return week mode and false (The calendar will target 'now' by default)
+            -If there is only one, return week mode and date of the start of the event.
+            -If there are several events entirely on the same week, return week mode and start of first event.
+            -Else, return month mode and the date of the start of first event as initial date. (If they are
+            on the same month, this will display that month and therefore show all of them, which is expected)
+
+            :param opportunity_id: id of the opportunity of which we want to see the meetings.
+            :return tuple mode, initial_date:
+                - mode: selected mode of the calendar view, 'week' or 'month'
+                - initial_date: date of the start of the first relevant meeting. The calendar will target that date.
+        """
+        meeting_results = self.env["calendar.event"].search_read([('opportunity_id', 'in', [opportunity_id])], ['start', 'stop', 'allday'])
+        if not meeting_results:
+            return "week", False
+
+        user_tz = self.env.user.tz or self.env.context.get('tz')
+        user_pytz = pytz.timezone(user_tz) if user_tz else pytz.utc
+
+        # meeting_dt will contain one tuple of datetimes per meeting : (Start, Stop) as per user time zone.
+        meeting_dt = []
+        now_dt = datetime.now().astimezone(user_pytz).replace(tzinfo=None)
+
+        # When creating an allday meeting, whatever the TZ, it will be stored the same e.g. 00.00.00->23.59.59 in utc or
+        # 08.00.00->18.00.00. Therefore we must not put it back in the user tz but take it raw.
+        for m in meeting_results:
+            if m.get('allday'):
+                meeting_dt.append((m.get('start'), m.get('stop')))
+            else:
+                meeting_dt.append((m.get('start').astimezone(user_pytz).replace(tzinfo=None),
+                                   m.get('stop').astimezone(user_pytz).replace(tzinfo=None)))
+
+        # If there are meetings that are still ongoing or to come, only take those.
+        unfinished_meeting_dt = [m for m in meeting_dt if m[1] >= now_dt]
+        relevant_meeting_dt = unfinished_meeting_dt if unfinished_meeting_dt else meeting_dt
+        n_relevant_meeting = len(relevant_meeting_dt)
+
+        if n_relevant_meeting == 1:
+            return "week", relevant_meeting_dt[0][0].date()
+        else:
+            # Range of meetings
+            earliest_start_dt = min(m[0] for m in relevant_meeting_dt)
+            latest_stop_dt = max(m[1] for m in relevant_meeting_dt)
+
+            # The week start day depends on language. We fetch the week_start of user's language. 1 is monday.
+            lang_week_start = self.env["res.lang"].search_read([('code', '=', self.env.user.lang)], ['week_start'])
+            week_start_index = int(lang_week_start[0].get('week_start') or '1')
```

Comment:
int(lang_week_start[0].get('week_start', '1'))
but is it not '0' instead of '1' ?
-> So the 8 in earliest_start_dt_weekday becomes 7 and it looks more logical, IMO

Reply:
In the res.lang table, week_start is between 1 and 7 :) so I substracted one to make it 0-6